### PR TITLE
Add support for Play 2.4

### PR DESCRIPTION
**Pull request by mkurz**
I squashed the commits from @vmouta from #249 into one nice, single commit.
Then I added lots of enhancements and cleaned up the code. Unfortunately I also squashed all the work I have done into vmouta's commit, but I don't really care as I don't have time to extract my changes out of his commit. I am sure there is some git magic to do so but I don't want to waste time.

This PR does not use dependency injection yet, it still relies on a `play.plugins` file. It's ok for now, but for Play 3.0 this has to be addressed, as Plugins will be removed in favor of Modules written with DI.

The release notes for deadbolt can be found [here](https://github.com/schaloner/deadbolt-2-java) and [here](https://groups.google.com/forum/?fromgroups#!searchin/play-framework/deadbolt%7Csort:date/play-framework/KZcCdbWZXvQ/m4nMOE6SqvkJ).

@joscha Could you please review this PR and if you are fine with it merge and release a snapshot so people can test?

### CODE REVIEW

**Comment by vmouta** [comment]
I would suggest also to update to bootstrap 3.3.4.

**Comment by mkurz** [comment]
@joscha Could you have a look at Travis build? Also my [last commit](https://github.com/mkurz/play-authenticate/commit/2d4efa199a6e466c6c510aab31788a132207f9ce) can be reverted as soon there are stable versions of easymail and deadbolt.

**Comment by mkurz** [comment]
@joscha If you need any help or infos let my know :smile:

File: samples/java/play-authenticate-usage/app/models/User.java [hunk 1]
```diff
@@ -173,7 +172,7 @@ public static User create(final AuthUser authUser) {
 		}
 
 		user.save();
-		user.saveManyToManyAssociations("roles");
+		// user.saveManyToManyAssociations("roles");
```
Comment (joscha) [inline samples/java/play-authenticate-usage/app/models/User.java]: @mkurz is this being done automagically now?

Comment (mkurz) [inline samples/java/play-authenticate-usage/app/models/User.java]: It seems so, the method `saveManyToManyAssociations` does not exist anymore. I was running the app and did some testing and it seems the role was saved correctly. E.g. I could not access the restricted demo page.

**Comment by Enalmada** [comment]
I could be wrong but when converting my java project to 2.4, I seem to remember changing ebean associations to something like this to get it to compile:
```Ebean.saveManyToManyAssociations(user, "roles")```

Though now I wish I just tried commenting it out.

**Comment by joscha** [comment]
it seems as if the build breaks because of the aggregation of the test-app and play authenticate.
This one does not seem to work any more:
```sbt
lazy val playAuthenticate = (project in file("modules/play-authenticate")).enablePlugins(PlayJava)

lazy val root = (project in file("."))
  // ...
  .dependsOn(playAuthenticate)
  .aggregate(playAuthenticate)
```

I read the [sbt docs](http://www.scala-sbt.org/0.13/tutorial/Multi-Project.html) but it appears to be correct. Any ideas?

**Comment by mkurz** [comment]
No idea.
For developing local on my machine I published a snapshot into my local ivy repos and then commented these lines out and instead added
```
  "com.feth"      %% "play-authenticate" % "0.7.0-SNAPSHOT",
```
into `appDependencies`.

**Comment by mkurz** [comment]
This approach of course doesn't work for us as there is no SNAPSHOT available yet.

Also, I predict that when Travis wants to run the tests for the `play-authenticate-usage` and the `play-authenticate-simple-oauth` it will fail because of the same reason (both depend on a `0.7.0-SNAPSHOT` in its `appDependencies`, but there is no `SNAPSHOT` published yet.

**Comment by mkurz** [comment]
Or wait - am I wrong? Because Travis of course will publish a SNAPSHOT into its local repos, so my approach from https://github.com/joscha/play-authenticate/pull/258#issuecomment-110825266 should work, shouldn't it?

**Comment by mkurz** [comment]
I added a commit - waiting for Travis. I can remove the commit if it's a bad idea/approach.

**Comment by joscha** [comment]
travis published locally, so that is not a problem. your commit should make the test-app initialise, but the tests should fail, because play-authenticate is missing.

**Comment by joscha** [comment]
@mkurz does the sample app work for you? I keep on getting:
```
[IllegalStateException: JSR-303 validated property 'repeatPassword' does not have a corresponding accessor for data binding - check your DataBinder's configuration (bean property versus direct field access)]
```

**Comment by mkurz** [comment]
Both worked for me. Which one do you mean? I can try again.

**Comment by mkurz** [comment]
I don't understand why Travis is failing. Why does it not find `/home/travis/.ivy2/local/com.feth/play-easymail_2.11/0.7.0-SNAPSHOT/` after `publish-local` has run before? Why is it working on my machine then?

**Comment by mkurz** [comment]
Man, I am stupid. It couldn't find `play-easymail` (not `play-authenticate`). I should buy a stronger pair of glasses...

**Comment by mkurz** [comment]
@joscha About the samples apps, make sure you deleted all your `target`, `bin` folders, etc. so you start from clean. When I upgraded apps to Play 2.4 sometimes there was something cached that didn't make it work.

**Comment by mkurz** [comment]
@joscha Both samples apps work fine for me...

**Comment by mkurz** [comment]
@joscha Build is green! :tada: I did some cleanup in the last commits. Any chance for a (snapshot) release today?

**Comment by joscha** [comment]
hmm - @mkurz - now that the aggregation and dependency is gone, that means it is using the dependency from the repository - so basically the tests don't run against the code that was checked in - I have been trying to fix this the last few hours, but didn't have any luck. Do you have an idea if the symlinks are the problem?

**Comment by mkurz** [comment]
@joscha I don't really know. I don't have very deep knowledge of sbt. I would have to dig into and try...

**Comment by mkurz** [comment]
Is there any difference in writing
```
(project in file("modules/play-authenticate"))
```
or
```
project.in(file("modules/play-authenticate"))
```
?

Also what about instead using symlinks using this (don't know hoch much `../` are really necassary):
```
file("../../play-authenticate")
```
?

**Comment by vmouta** [comment]
Saw some people using the ./ - probably it doesn’t make any diference but not in condition to try now.
(project in file("./modules/play-authenticate"))

**Comment by joscha** [comment]
@mkurz they are equivalent according to the manual. Using `../`  is not possible as you can not leave the root project's base folder unfortunately. I also tried copying the code into the modules folder just to try, but it didn't seem to change much.

**Comment by joscha** [comment]
figured it out I think. Will create new PR and try to generate a snapshot...

**Comment by mkurz** [comment]
Nice!

**Comment by joscha** [comment]
see #259.

**Comment by mkurz** [comment]
Closing in favor of #259

**Pull request closed without merging**